### PR TITLE
#infra Keychain SecItem migration steps 0-1: design plan + characterization suite

### PR DESCRIPTION
**Pull request by Kaspik**
## Summary

First PR of the SPKeychain → `SecItem*` migration (the warnings plan's highest-risk deferred item). This lands the design plan plus Steps 0–1: the keychain test harness and a characterization suite pinning the legacy `SecKeychain*` implementation exactly as it behaves — defects included — before anything changes. Steps 0 and 1 share this PR because the harness only exists for these tests; every later step is its own PR.

- **`docs/development/keychain-secitem-migration-plan.md`** — the full design: characterize-first, stay on the file-based login keychain (zero data migration), retire the tunnel assistant's direct keychain read (which also answers the XPC plan's open question 3), then a Swift `SAKeychain` proven by a legacy↔Swift cross-compatibility matrix. Cross-linked from the warnings and modernization plans.
- **`SAKeychainProviding`** — public `@objc` Swift protocol mirroring SPKeychain's selectors, adopted by SPKeychain via a class extension. This is how Swift tests reach SPKeychain without giving the test target a bridging header (the documented B2b sharp edge): `NSClassFromString` + a real runtime conformance check. It is also the seam the future `SAKeychain` implements, so the identical suite reruns against both.
- **`SAKeychainTestSupport`** — keychain tests run against the real login keychain under a unique per-run `Sequel Ace Test <token>` namespace, with a defensive sweep (aborted-run leftovers included) and a raw-`SecItem` availability probe that skips the store suite where the keychain is locked — deliberately probing without SPKeychain so its modal error alerts can never hang the runner.
- **33 characterization tests**: the four persisted name/account format strings byte-exact (the keychain wire format), CRUD round-trips, add-is-a-silent-no-op-when-present, persisted item shape (service/account/label/`kSecAttrGeneric`), unicode/empty/long payloads, nil rejection, both update recovery branches (−25300 → fresh add, −25299 → replace destination), and the label-stays-behind-on-rename quirk.

**Confirmed live, previously only suspected**: `updateItemWithName:account:toPassword:` scrambles its arguments — the item is renamed to the *password*, its account becomes the *name*, and the stored secret becomes the *account*. No production callers (all four call sites use the five-argument form); the pinning test documents it and Step 2 deletes it.

## Warnings accounting

SPKeychain.m now also compiles into the Unit Tests target, so its 10 deprecation warnings appear a third time (52 → 62 raw on the test scheme; **unique lines unchanged at 20**). Temporary — Step 5 deletes the file and the whole block goes to zero.

## Test plan

- [x] Full "Unit Tests" scheme: 1178 tests, 0 failures (63 pre-existing env-gated skips); both new suites ran (not skipped)
- [x] Clean `build-for-testing`: 0 errors, no new unique warnings
- [x] Login keychain verified free of test items after the run (namespace sweep works)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Added a shared keychain interface for consistent credential storage operations.
  * Preserved existing credential formats and behavior while preparing for modern keychain storage.

* **Tests**
  * Added coverage for naming, account formatting, CRUD operations, metadata, Unicode, boundary cases, and updates.
  * Added utilities to safely check keychain availability and clean up test data.

* **Documentation**
  * Added a migration roadmap for modernizing keychain storage.
  * Updated modernization and warnings plans to reflect the migration status.

### CODE REVIEW

File: UnitTests/SAKeychainStoreCharacterizationTests.swift [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+    }
+
+    // MARK: - Persisted item shape
+
+    func testAddStoresServiceAccountLabelAndGenericAttribute() throws {
+        let name = service("item"), account = "user@host/db"
+        store.add(password: "pw", name: name, account: account)
+
+        let attrs = try XCTUnwrap(SAKeychainTestSupport.attributes(service: name, account: account))
+        XCTAssertEqual(attrs[kSecAttrService as String] as? String, name)
```
[the remark this answers is not in the document]
Comment (Kaspik) [reply UnitTests/SAKeychainStoreCharacterizationTests.swift:82]: Fixed in c13c0061f — explicit `import Security` added (it compiled via transitive resolution, but the explicit import is right).

File: UnitTests/SAKeychainTestSupport.swift [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+            kSecClass as String: kSecClassGenericPassword,
+            kSecMatchLimit as String: kSecMatchLimitAll,
+            kSecReturnAttributes as String: true,
+        ]
+        var result: CFTypeRef?
+        guard SecItemCopyMatching(query as CFDictionary, &result) == errSecSuccess,
+              let items = result as? [[String: Any]] else { return }
+        for item in items {
+            guard let service = item[kSecAttrService as String] as? String,
+                  service.hasPrefix(testServicePrefix) else { continue }
```
[the remark this answers is not in the document]
Comment (Kaspik) [reply UnitTests/SAKeychainTestSupport.swift:122]: Fixed in c13c0061f — the namespace is now `Sequel Ace Unit Test 8C41F2D6`, a fixed random token that keeps the aborted-run sweep working while making third-party collision implausible.

File: Source/Other/Keychain/SAKeychainProviding.swift [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+/// Public rather than internal so it appears in every target's generated
+/// `sequel-ace-Swift.h` — `SPKeychain.m` compiles into the app, the
+/// SequelAceTunnelAssistant, and (for the characterization suite) the Unit
+/// Tests target, and only public declarations are emitted into the generated
+/// header for targets that have no Objective-C bridging header.
+@objc public protocol SAKeychainProviding {
+
+    /// Adds a password with the item's label defaulting to `name`.
+    /// Silently does nothing when an item for (`name`, `account`) already
+    /// exists, or when any argument is nil/empty.
```
[the remark this answers is not in the document]
Comment (Kaspik) [reply Source/Other/Keychain/SAKeychainProviding.swift]: Fixed in c13c0061f — the doc now states the real contract: name/account nil-or-empty reject, only a nil password rejects, and an empty password is stored (round-trips as ""), matching the characterization test.

File: docs/development/keychain-secitem-migration-plan.md [hunk 4]
```diff
@@ -0,0 +1,396 @@
...
+> own design"). Siblings: `warnings-elimination-plan.md` (this project owns 10
+> of its remaining 20 warning lines), `ssh-tunnel-xpc-migration-plan.md` (open
+> question 3 there asks for the sequencing decision made here),
+> `modernization-followup-plan.md`, ground rules in `AGENTS.md`.
+
+## Why this is the highest-risk file in the codebase
+
+`Source/Other/Keychain/SPKeychain.m` (387 lines, unchanged in shape since the
+Sequel Pro era, one method already on `SecItem*`) is the sole owner of every
+saved MySQL and SSH password. It has **zero tests**. Its wire format — the
```
[the remark this answers is not in the document]
Comment (Kaspik) [reply docs/development/keychain-secitem-migration-plan.md]: Fixed in c13c0061f — now reads "had zero tests before this plan's Step 1".

File: docs/development/keychain-secitem-migration-plan.md [hunk 5]
```diff
@@ -0,0 +1,396 @@
...
+- `getPasswordForName:` returns nil (not empty) for a missing item.
+- `init` returns **nil** when `LIBMYSQL_ENABLE_CLEARTEXT_PLUGIN` is set
+  (issue #2437) — keychain access disabled wholesale.
+- All name/account inputs are validated as non-nil + non-empty
+  (`isValidName:acount:` [sic]); helpers return nil on invalid input.
+
+## Known and suspected defects (to be pinned or fixed in Step 1/2)
+
+Found by reading; Step 1's characterization tests confirm each before Step 2
+fixes it. Every fix is its own commit with a test that fails first.
```
[the remark this answers is not in the document]
Comment (Kaspik) [reply docs/development/keychain-secitem-migration-plan.md:71]: Fixed in c13c0061f — the Step 1 intro now says every *testable* defect is confirmed, with the nil-password crasher explicitly called out as asserted post-fix in Step 2.

File: docs/development/keychain-secitem-migration-plan.md [hunk 6]
```diff
@@ -0,0 +1,396 @@
...
+  this is prompt-free locally and on CI (GitHub Actions runners have an
+  unlocked login keychain; a `SecKeychainGetStatus`-free probe — write one
+  item, read it back — gates the suite with `XCTSkip` on any machine where
+  the keychain is locked or unavailable, e.g. some SSH sessions).
+- The suite must also pass with `LIBMYSQL_ENABLE_CLEARTEXT_PLUGIN` **unset**
+  — the init guard means these tests construct the store *after* asserting
+  the env var is absent, and one dedicated test covers the guard itself.
+- These tests live in a new `UnitTests/SAKeychainStoreTests.swift` (naming
+  says what it is; "integration-ish" unit tests). Pure-logic tests (naming,
+  validation) go in ordinary suites and never touch the keychain.
```
[the remark this answers is not in the document]
Comment (Kaspik) [reply docs/development/keychain-secitem-migration-plan.md]: Fixed in c13c0061f — the plan now names `UnitTests/SAKeychainStoreCharacterizationTests.swift`.

File: docs/development/modernization-followup-plan.md [hunk 7]
```diff
@@ -6,7 +6,9 @@
 > `docs/development/help-viewer-rewrite-plan.md` (executed), the platform floor
 > in `docs/development/macos-13-minimum-plan.md` (executed, merged as #2587),
 > the SSH-tunnel IPC design in `docs/development/ssh-tunnel-xpc-migration-plan.md`
-> (not started), agent ground rules in `AGENTS.md`.
+> (not started), the keychain `SecItem*` design in
+> `docs/development/keychain-secitem-migration-plan.md` (not started), agent
+> ground rules in `AGENTS.md`.
```
[the remark this answers is not in the document]
Comment (Kaspik) [reply docs/development/modernization-followup-plan.md]: Fixed in c13c0061f — the index now separates design status from execution ("design complete; execution starting with its Steps 0-1"). Later PRs in the stack advance this label as the steps land.

File: docs/development/warnings-elimination-plan.md [hunk 8]
```diff
@@ -474,8 +474,10 @@ deferred projects below.
 
 - **SPKeychain SecKeychain* API (~15)** — migrate to `SecItem*` with in-place
   migration of users' stored passwords. Highest risk item in the codebase;
-  needs its own design (and pairs with the NSConnection item below since the
-  tunnel assistant reads passwords).
+  **now designed**: see `docs/development/keychain-secitem-migration-plan.md`
+  (characterize-first, stays on the file-based login keychain so no data
+  migration, retires the tunnel assistant's direct keychain read — which also
```
[the remark this answers is not in the document]
Comment (Kaspik) [reply docs/development/warnings-elimination-plan.md:480]: Fixed in c13c0061f — reworded to "migrate the keychain *code* to SecItem* in place; users' stored password records are retained as-is (no record migration)".

File: Source/Other/Keychain/SAKeychainProviding.swift [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+    func deletePassword(name: String?, account: String?)
+
+    @objc(passwordExistsForName:account:)
+    func passwordExists(name: String?, account: String?) -> Bool
+
+    /// Renames an item and replaces its password in place (preserving its
+    /// access list). Falls back to add when the source item is missing, and
+    /// replaces the destination when it already exists.
+    @objc(updateItemWithName:account:toName:account:password:)
+    func updateItem(name: String?, account: String?, toName newName: String?, newAccount: String?, password: String?)
```
[the remark this answers is not in the document]
Comment (Kaspik) [reply Source/Other/Keychain/SAKeychainProviding.swift:76]: Deliberately not changed: the parameter stays nullable to mirror the annotated legacy header, and the strlen(NULL) crash it can reach is defect 2 of the migration plan — fixed with a guard and a test in the next PR of this stack (#2612, commit c1cc59c57), and guarded in the SecItem* replacement too (#2614). Making it nonoptional here would change the ObjC-imported surface mid-characterization.

**Review by Jason-Morcos — APPROVED** [review]
